graspologic/cluster/rclust.py

```diff
@@ -0,0 +1,377 @@
+    graspy.cluster.AutoGMMCluster
+
+    References
+    ----------
+    .. [1]  Lyzinski, V., Tang, M., Athreya, A., Park, Y., & Priebe, C. E
+            (2016). Community detection and classification in hierarchical
+            stochastic blockmodels. IEEE Transactions on Network Science and
+            Engineering, 4(1), 13-26.
+    """
+
+    def __init__(
+        self,
+        selection_criteria=None,
+        cluster_method="GMM",
+        parent=None,
+        min_components=1,
+        max_components=10,
+        cluster_kws={},
+        min_split=1,
+        max_level=20,
+        delta_criter=None,
+        likelihood_ratio=None,
+    ):
+
+        _check_common_inputs(
+            min_components, max_components, cluster_kws, embed_kws=None
+        )
+
+        if cluster_method not in ["GMM", "KMeans", "Spherical-KMeans"]:
+            msg = "clustering method must be one of"
+            msg += "{GMM, Kmeans, Spherical-KMeans}"
+            raise ValueError(msg)
+
+        if delta_criter is not None:
+            if delta_criter <= 0:
+                raise ValueError("delta_criter must be positive")
+
+        if likelihood_ratio is not None:
+            if likelihood_ratio <= 0 or likelihood_ratio >= 1:
+                raise ValueError("likelihood_ratio must be in (0,1)")
+
+        self.parent = parent
+        self.min_components = min_components
+        self.max_components = max_components
+        self.cluster_method = cluster_method
+        self.min_split = min_split
+        self.selection_criteria = selection_criteria
+        self.cluster_kws = cluster_kws
+        self.min_split = min_split
+        self.max_level = max_level
+        self.delta_criter = delta_criter
+        self.likelihood_ratio = likelihood_ratio
+
+    def fit(self, X):
+        """
+        Fits clustering models to the data as well as resulting clusters
+
+        Parameters
+        ----------
+        X : array_like, shape (n_samples, n_features)
+
+        Returns
+        -------
+        self : object
+            Returns an instance of self.
+        """
+        self.fit_predict(X)
+        return self
+
+    def fit_predict(self, X, level=None):
+        """
+        Fits clustering models to the data as well as resulting clusters
+        and using fitted models to predict a hierarchy of labels
+
+        Parameters
+        ----------
+        X : array_like, shape (n_samples, n_features)
+        level: int, optional (default=None)
+
+        Returns
+        -------
+        labels : array_label, shape (n_samples, n_levels)
+            if no level specified; otherwise, shape (n_samples,)
+        """
+        X = check_array(X, dtype=[np.float64, np.float32], ensure_min_samples=1)
+        if level is not None:
+            if not isinstance(level, int):
+                raise TypeError("level must be an int")
+            elif level < 1:
+                raise ValueError("level must be positive")
+
+        self.X = X
```

Comment:
apologies if this seems nitpicky but in general models don't store the data. it could be massive. and I think we can circumvent using the same recursive structure for `predict`